sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2739,6 +2752,34 @@ class Dataset[T] private[sql](
   }
 
   /**
+   * Returns true when the Dataset is cached and materialized.
+   *
+   * @group basic
+   * @since 2.3.0
+   */
+  def isMaterialized(): Boolean = {
+    queryExecution.sparkPlan match {
+      case i: InMemoryTableScanExec =>
+        val blockManager = sparkSession.sparkContext.env.blockManager
+
+        val rdd = i.relation.cachedColumnBuffers
+        val blockIDs = rdd.partitions.indices.map(index => RDDBlockId(rdd.id, index))
+
+        var foundNonexistentBlocks = false
+        blockIDs.foreach { bid =>
+          if (blockManager.get(bid).isEmpty) {
+            foundNonexistentBlocks = true
+          } else {
+            blockManager.releaseLock(bid)
```

Comment:
Do you not need to release the block in all cases? maybe in try-finally? I don't know the BlockManager well though.